.evergreen/config.yml

```diff
@@ -11,7 +11,7 @@ command_type: system
 # Protect ourself against rogue test case, or curl gone wild, that runs forever
 # Good rule of thumb: the averageish length a task takes, times 5
 # That roughly accounts for variable system performance for various buildvariants
-exec_timeout_secs: 1800 # 30 minutes is the longest we'll ever run
+exec_timeout_secs: 3600 # 60 minutes is the longest we'll ever run
```

Comment:
Now that libmongoc is using the streaming protocol in 4.4+, the failpoint based spec tests take a lot longer (see SERVER-45686 and SERVER-50500). Unfortunately, the 4.4+ auth sharded tests now run over the 30 minute mark and fail due to timing out even with sub-500ms minHeartbeatFrequency, necessitating this change.

From looking at the evergreen runs of the C driver, I couldn't find a 4.4+ sharded auth+ssl variant to compare against. I also checked the Go driver's evergreen runs, and their 4.4+ sharded auth+ssl tests only take about 6m, which seems impossible (?). I will do some more digging tomorrow to see if our tests are slower than the norm.

Reply:
After doing some digging, I found that the deinitiliziar of a sync client takes about 0.5s to complete every time, and in the spec test runner we were creating a lot of short-lived clients for things like fail points or verifying the test outcomes. We were also running `distinct` on every shard for every test, even when it wasn't totally necessary. I made some changes to stop the extra `distinct` calls and also to try to reuse the clients as much as possible and was able to get significant performance improvements.

On a 4.4 sharded cluster, I was able to actually make some decent improvements even on the test times we're seeing on master, particularly for retryable reads:

|                  | master | 1.17 first pass | 1.17 after optimization |   |
|------------------|--------|-----------------|-------------------------|---|
| Transactions     | 233s   | timeout         | 212s                    |   |
| Retryable Writes | 16s    | 88s             | 19s                     |   |
| Retryable Reads  | 150s   | 558s            | 42s                     |   |

The optimized column are the tests running on my machine, so they'll probably be a bit slower on evergreen.

There's still a lot to be improved on, so I filed SWIFT-1008 to track work for further improving the test suite.

Reply:
Oh here's where I can comment :smile: In order to make those tests run faster you need to reduce your `heartbeatFrequencyMS`. I can't find any reference to this in the specs, and it seems like a major oversight if we forgot to add it, but that value is what's used for the `maxAwaitTimeMS` of the call to stream "hellos" so you need to set it a relatively small values in tests where you expect failure in order to speed up cancellation of those streams.

Reply:
I think it's hard to see this because the PR is quite large and Patrick already did some iterating on the initial work, but he already adjusted `heartbeatFrequencyMS` down significantly for tests at the start, and we were still experiencing the timeouts, what he describes here is follow-on work.

regarding the clients, that seems weird. this program (using the async API directly, but doing exactly what the sync client would be doing in its shutdown) outputs on average ~0.002s for me locally:
```swift
let elg = MultiThreadedEventLoopGroup(numberOfThreads: 4)
let client = try MongoClient(using: elg)
let _ = try client.db("test").runCommand(["isMaster": 1]).wait()

let start = Date()
try client.syncClose()
try elg.syncShutdownGracefully()
print("time: \(Date().timeIntervalSince(start))")
```

Reply:
Hm, I ran your example and got the similar ~0.5s time. Almost all the time seems to be spent in the call to `mongoc_client_pool_destroy`. Are you using the 1.17.0 branch?